### PR TITLE
Add support for binary data extension protocol and FP16 datatype

### LINKED ISSUE #3643: Inference gRPC/Rest client to support FP16
/kind feature

**Describe the solution you'd like**
Triton client supports FP16, kserve rest/grpc should also support transporting FP16.
https://github.com/kserve/kserve/blob/master/python/kserve/kserve/protocol/infer_type.py#L278-L296

**Anything else you would like to add:**
[Miscellaneous information that will assist in solving the issue.]

**Links to the design documents:**
[Optional, start with the short-form [RFC template](https://docs.google.com/document/d/1UcBeLfZ_JMGpVrPJmYtEIVH_9Y4U3AEKQdq_IKuOMrU) to outline your ideas and get early feedback.]
[Required, use the longer-form [design doc template](https://docs.google.com/document/d/1Mtoui_PP2a9N59NjYHnsvrdJ8t2iKFwIJAx1zRO_I1c) to specify and discuss your design in more detail]

### PR DESCRIPTION
**Pull request by sivanantha321**
**What this PR does / why we need it**:

**Which issue(s) this PR fixes** *(optional, in `fixes #<issue number>(, fixes #<issue_number>, ...)` format, will close the issue(s) when PR gets merged)*:
Fixes #3643

https://github.com/triton-inference-server/server/blob/main/docs/protocol/extension_binary_data.md

**Type of changes**
Please delete options that are not relevant.

- [ ] Bug fix (non-breaking change which fixes an issue)
- [x] New feature (non-breaking change which adds functionality)
- [ ] Breaking change (fix or feature that would cause existing functionality to not work as expected)
- [ ] This change requires a documentation update

**Feature/Issue validation/testing**:

Please describe the tests that you ran to verify your changes and relevant result summary. Provide instructions so it can be reproduced.
Please also list any relevant details for your test configuration.

- [ ] Test A
- [ ] Test B

- Logs

**Special notes for your reviewer**:

1. Please confirm that if this PR changes any image versions, then that's the sole change this PR makes.

**Checklist**:

- [x] Have you added unit/e2e tests that prove your fix is effective or that this feature works?
- [x] Has code been commented, particularly in hard-to-understand areas?
- [ ] Have you made corresponding changes to the documentation?

**Release note**:

```release-note

```

**Re-running failed tests**

- `/rerun-all` - rerun all failed workflows.
- `/rerun-workflow <workflow name>` - rerun a specific failed workflow. Only one workflow name can be specified. Multiple /rerun-workflow commands are allowed per comment.

### CODE REVIEW

File: python/kserve/kserve/model.py [hunk 1]
```diff
@@ -345,8 +340,13 @@ async def _http_predict(
             if "x-b3-traceid" in headers:
                 predict_headers["x-b3-traceid"] = headers["x-b3-traceid"]
         if isinstance(payload, InferRequest):
-            payload = payload.to_rest()
-        data = orjson.dumps(payload)
+            payload, json_length = payload.to_rest()
+            if json_length is not None:
```
Comment (yuzisun) [inline python/kserve/kserve/model.py]:
```suggestion
            if json_length:
```

File: python/kserve/kserve/protocol/dataplane.py [hunk 2]
```diff
@@ -224,11 +227,22 @@ def model_ready(self, model_name: str) -> bool:
...
+        body,
+        headers,
+        protocol_version: str = PredictorProtocol.REST_V1.value,
+        model_name: str = None,
+    ) -> Tuple[Union[Dict, InferRequest], Dict]:
         t1 = time.time()
         attributes = {}
         if isinstance(body, InferRequest):
             return body, attributes
+        elif isinstance(body, InferenceRequest) or (
```
Comment (yuzisun) [inline python/kserve/kserve/protocol/dataplane.py:241]: I remember `InferenceRequest` was always converted to `InferRequest` before hitting here

File: python/kserve/kserve/protocol/rest/server.py [hunk 3]
```diff
@@ -174,7 +173,7 @@ def create_application(self) -> FastAPI:
                     r"/v2/models/{model_name}/infer",
                     v2_endpoints.infer,
                     methods=["POST"],
-                    response_model=InferenceResponse,
+                    response_model=None,
```
Comment (yuzisun) [inline python/kserve/kserve/protocol/rest/server.py]: The downside of this is that we lose the validation when returning back the `InferenceResponse`

File: test/e2e/predictor/test_sklearn.py [hunk 4]
```diff
@@ -221,6 +223,32 @@ def test_sklearn_v2():
     res = predict(service_name, "./data/iris_input_v2.json", protocol_version="v2")
     assert res["outputs"][0]["data"] == [1, 1]
 
+    raw_res = predict(
+        service_name,
+        "./data/iris_input_v2_binary.json",
```
Comment (yuzisun) [inline test/e2e/predictor/test_sklearn.py:256]: How do we test sending inputs with binary data for REST ?

**Comment by yuzisun** [comment]
@sivanantha321 Need a rebase after merging the inference client PR

File: python/kserve/kserve/model.py (hunk 1 above)
Comment (sivanantha321) [reply python/kserve/kserve/model.py]: Resolved

File: test/e2e/predictor/test_sklearn.py (hunk 4 above)
Comment (sivanantha321) [reply test/e2e/predictor/test_sklearn.py:256]: Added in custom transformer test and also included in python tests

File: python/kserve/kserve/protocol/rest/server.py (hunk 3 above)
Comment (sivanantha321) [reply python/kserve/kserve/protocol/rest/server.py]: The response will be validated [here](https://github.com/kserve/kserve/pull/3685/files#diff-3aa3dd39255bdc2c68b576b8abdc038a614324fd09542b3d137ac9c28e0d8f2bR180)

File: python/kserve/kserve/protocol/dataplane.py [hunk 5]
```diff
@@ -280,15 +288,35 @@ def decode_cloudevent(self, body) -> Tuple[Union[Dict, InferRequest], Dict]:
...
 
+    def decode_inference_request(
+        self, body: Union[bytes, InferenceRequest], headers: Dict, model_name: str
+    ) -> InferRequest:
+        if isinstance(body, bytes):
+            json_length = headers.get(INFERENCE_CONTENT_LENGTH_HEADER, None)
+            if json_length is None:
+                raise InvalidInput(
+                    f"A byte input is received. But, the"
+                    f"'{INFERENCE_CONTENT_LENGTH_HEADER}' header is missing."
```
Comment (yuzisun) [inline python/kserve/kserve/protocol/dataplane.py]:
```suggestion
                    f"received byte inputs, but the"
                    f"'{INFERENCE_CONTENT_LENGTH_HEADER}' header is missing."
```

File: test/e2e/common/utils.py [hunk 6]
```diff
@@ -51,13 +51,14 @@ def grpc_client(host):
...
         ],
+        timeout=120,
     )
 
 
 async def predict_isvc(
     client: InferenceRESTClient,
     service_name,
-    input_path,
+    input: [str, InferRequest],
```
Comment (yuzisun) [inline test/e2e/common/utils.py]:
```suggestion
    input: Union[str, InferRequest],
```

File: test/e2e/common/utils.py [hunk 7]
```diff
@@ -89,13 +90,16 @@ async def predict(
     client: InferenceRESTClient,
     url,
     host,
-    input_path,
+    input: [str, InferRequest],
```
Comment (yuzisun) [inline test/e2e/common/utils.py]:
```suggestion
    input: Union[str, InferRequest],
```

File: python/kserve/test/test_server.py [hunk 8]
```diff
@@ -268,6 +272,72 @@ def __init__(self, name):
...
+                .as_numpy()
+                .astype(np.float16)
+                .flatten(),
+                "fp32_output": request.get_input_by_name("fp32_input")
+                .as_numpy()
+                .flatten(),
+            }
+        )
+        # Fixme: Gets only the 1st element of the input
+        # inputs = get_predict_input(request)
```
Comment (yuzisun) [inline python/kserve/test/test_server.py:297]:
```suggestion
```

File: python/kserve/test/test_server.py [hunk 9]
```diff
@@ -268,6 +272,72 @@ def __init__(self, name):
...
+                .as_numpy()
+                .flatten(),
+                "fp32_output": request.get_input_by_name("fp16_input")
+                .as_numpy()
+                .astype(np.float32)
+                .flatten(),
+            }
+        )
+        # Fixme: Gets only the 1st element of the input
+        # inputs = get_predict_input(request)
```
Comment (yuzisun) [inline python/kserve/test/test_server.py:330]:
```suggestion
```

**Review by yuzisun — APPROVED** [review]
/lgtm
/approve

**Pull request merged**